### PR TITLE
Code Refactor

### PR DESCRIPTION
**Pull request by pirouet**
Modify two methods to improve code readability and update to one return method. Third class is purely updating the use statements

### CODE REVIEW

**Comment by ceeram** [comment]
i personally prefer the original code with early returns, the happy path :)

File: src/Console/Shell.php [hunk 1]
```diff
@@ -265,18 +265,16 @@ public function hasTask($task)
...
     {
+        $hasMethod = false;
         try {
-            $method = new \ReflectionMethod($this, $name);
-            if (!$method->isPublic()) {
-                return false;
-            }
-            if ($method->getDeclaringClass()->name === 'Cake\Console\Shell') {
-                return false;
+            $method = new ReflectionMethod($this, $name);
```
Comment (ceeram) [inline src/Console/Shell.php]: this line causes tests to fail, as no use statement was added on top

File: src/Console/Shell.php [hunk 2]
```diff
@@ -265,18 +265,16 @@ public function hasTask($task)
...
+            if($method->isPublic() && $method->getDeclaringClass()->name === 'Cake\Console\Shell') {
+                $hasMethod = true;
             }
-            return true;
-        } catch (\ReflectionException $e) {
-            return false;
+        } catch (ReflectionException $ex) {
```
Comment (ceeram) [inline src/Console/Shell.php]: also requires use statement if leading backslash is being removed

Comment (pirouet) [inline src/Console/Shell.php]: Damn! Overshoot on my part. Adding it now in another commit

File: src/Console/Shell.php [hunk 3]
```diff
@@ -265,18 +267,16 @@ public function hasTask($task)
...
+        $hasMethod = false;
         try {
-            $method = new \ReflectionMethod($this, $name);
-            if (!$method->isPublic()) {
-                return false;
-            }
-            if ($method->getDeclaringClass()->name === 'Cake\Console\Shell') {
-                return false;
+            $method = new ReflectionMethod($this, $name);
+            if($method->isPublic() && $method->getDeclaringClass()->name === 'Cake\Console\Shell') {
```
Comment (josegonzalez) [inline src/Console/Shell.php]: I think this is missing a space after the `if`.

Comment (pirouet) [inline src/Console/Shell.php]: space added @josegonzalez  ... I'll get there eventually on this PR ;)

File: src/Console/Shell.php [hunk 4]
```diff
@@ -265,18 +267,16 @@ public function hasTask($task)
...
-                return false;
+            $method = new ReflectionMethod($this, $name);
+            if ($method->isPublic() && $method->getDeclaringClass()->name === 'Cake\Console\Shell') {
+                $hasMethod = true;
             }
-            return true;
-        } catch (\ReflectionException $e) {
-            return false;
+        } catch (ReflectionException $ex) {
+            $this->log($ex->getMessage());
```
Comment (markstory) [inline src/Console/Shell.php]: Why are we logging here?

File: src/Controller/Controller.php [hunk 5]
```diff
@@ -646,18 +646,16 @@ public function paginate($object = null)
...
-        if (!$method->isPublic()) {
-            return false;
-        }
-        if ($method->getDeclaringClass()->name === 'Cake\Controller\Controller') {
-            return false;
+            if($method->isPublic() && $method->getDeclaringClass()->name === 'Cake\Controller\Controller') {
+                $isAction = true;
+            }
+        } catch (ReflectionException $ex) {
+            $this->log($ex->getMessage());
```
Comment (markstory) [inline src/Controller/Controller.php]: We weren't logging before. This will effectively log every 404.

**Comment by pirouet** [comment]
@markstory the idea was to give a bit of information in the event of an exception. Happy to take out though

**Comment by markstory** [comment]
@pirouet The issue is that the 404's will be logged elsewhere, and by adding additional logging at a non-debug level we can easily overwhelm a large application's logging infrastructure.

**Comment by dakota** [comment]
I find the proposed changes to be worse in terms of readability. I'm a big fan of returning early and single indentation level (as per the Object Calisthenics rules).

Some more information:
* http://williamdurand.fr/2013/06/03/object-calisthenics/#1-only-one-level-of-indentation-per-method
* http://www.slideshare.net/rdohms/bettercode-phpbenelux212alternate
* https://www.youtube.com/watch?v=GtB5DAfOWMQ

**Comment by markstory** [comment]
I'm ok with merging these changes if the log() statements are removed and tests pass.

**Comment by pirouet** [comment]
I will remove the logging from the two methods within the next hour or two, when I can get to my laptop

**Comment by pirouet** [comment]
@markstory So, it appears the tests keep erroring; having a look at this I cannot see why (the code should still return false as per the current code. As this is the case, i'm happy to close this PR and have another look to stop the builds from failing unless you have any advice as to why this is the case?

**Comment by dereuromark** [comment]
What is the benefit of the variables vs the previous early returns?

**Comment by jippi** [comment]
:-1: to changes that break the return early schema we try to follow :)

**Comment by ravage84** [comment]
:-1: Liked the previous approach more but :+1:  to the imports.

**Comment by pirouet** [comment]
Without sounding defence in any kind of way the idea was to combine the IF statement so it was easier to see the two conditions required to satisfy the requirements and use the single return method. My style of coding is to avoid using multiple returns which I know is down to a preference of style and not to everyones taste .

I will close this PR as it looks like the majority prefer the previous way and will only update the use statements.

**Pull request closed without merging**